Assets/Mirror/Components/NetworkSceneChecker.cs

```diff
@@ -0,0 +1,119 @@
+using UnityEngine;
+using System.Collections.Generic;
+using UnityEngine.SceneManagement;
+
+namespace Mirror
+{
+    /// <summary>
+    /// Component that controls visibility of networked objects between scenes.
+    /// <para>Any object with this component on it will only be visible to other objects in the same scene</para>
+    /// <para>This would be used when the server has multiple additive subscenes loaded to isolate players to their respective subscenes</para>
+    /// </summary>
+    [AddComponentMenu("Network/NetworkSceneChecker")]
+    [RequireComponent(typeof(NetworkIdentity))]
+    [HelpURL("https://mirror-networking.com/docs/Components/NetworkSceneChecker.html")]
+    public class NetworkSceneChecker : NetworkBehaviour
+    {
+        /// <summary>
+        /// Flag to force this object to be hidden from all observers.
+        /// <para>If this object is a player object, it will not be hidden for that client.</para>
+        /// </summary>
+        [Tooltip("Enable to force this object to be hidden from all observers.")]
+        public bool forceHidden;
+
+        static Dictionary<string, HashSet<NetworkIdentity>> sceneCheckerObjects;
+
+        [Scene, SerializeField]
+        string currentScene;
+
+        [ServerCallback]
+        void Awake()
+        {
+            if (sceneCheckerObjects == null)
+                sceneCheckerObjects = new Dictionary<string, HashSet<NetworkIdentity>>();
+
+            currentScene = gameObject.scene.name;
+        }
+
+        public override void OnStartServer()
+        {
+            if (!sceneCheckerObjects.ContainsKey(currentScene))
+                sceneCheckerObjects.Add(currentScene, new HashSet<NetworkIdentity>());
+
+            sceneCheckerObjects[currentScene].Add(netIdentity);
+        }
+
+        [ServerCallback]
+        void Update()
+        {
+            if (currentScene == gameObject.scene.name)
+                return;
+
+            // This object is in a new scene so observers in the prior scene
+            // and the new scene need to rebuild their respective observers lists.
+
+            // Remove this object from the hashset of the scene it just left
+            sceneCheckerObjects[currentScene].Remove(netIdentity);
+
+            // RebuildObservers of all NetworkIdentity's in the scene this object just left
+            RebuildSceneObservers();
```

Comment:
why is this in here?
each networkscenechecker component would rebuild observers for every one in sceneCheckerObjects unless I am missing something.
rebuildobservers is automatically called by unity already

Reply:
it's a private method instead of repeating the same for loop twice.

Reply:
@vis2k rebuild observers is not called automatically by unity,  we call it using a coroutine every few seconds with NetworkProximityChecker.

Basically what he is doing here is:  if this object changes scene,  notify all observers in the new scene,  which makes perfect sense.

otherwise,  we could put a coroutine in every object that checks the scene all the time, but that is more complicated and wasteful, since you only need to rebuild the observers when something enters or leaves the scene

Reply:
I still don't understand why this is called for each object.
what's the big picture here.
shouldn't each object just worry about it's own entry in the dictionary? why update everyone else too, when they get update calls themselves?

Reply:
It's actually only called for objects with this component, and the foreach loop only touches objects in the hashset of one dictionary entry and that hashset only has objects with this component.

Reply:
resolved in Discord via lengthy discussion.